### DISCUSSION TITLE
Few ideas for QoL improvement: manual sorting, tags, compile messages, and additional achieves details

### DISCUSSION THREAD

**Discussion by andrey-pavlov-13** [EddieDover/fvtt-player-achievements#48] [Ideas]
1. It would be great if when players get same achieve at the same time, it`s will not show few duplicated messages with different character name, but compile those in one message where will be written who got an achievement.
2. Additional item editing: 
  2.1 Adding possibility to edit text with different styles and add breaks(paragraphs). So GM can add extra details like quotes, jokes, notes etc.
  2.2 Allow GM a possibility to add some different color borders to icons, so player understand how rare each achievement. This one could easily replaced with tag system (i'll write about it below)
  2.3 Tag system could improve achievement editing and searching a lot. If GM have a bunch of items and don't remember how actually he named a achieve, searching by tag could really help him(of course if he added tag to achieve xD)
3. Extra ways for sorting achievements could really help Players and GM. Especially some manual sorting, so GM could group items in comfortable way. Because right now only way to 'manual' sort it`s editing code, what can end up with some errors if someone missed brakes, braces etc.

**Comment by EddieDover** [comment]
1. Entertaining this idea would require a refactor of some functionality related to how we display messages.  While it's not impossible, it would absolutely need offline messages to be finished first; otherwise I'd have to circle back around and modify the code after, leading to more possible instabilities. 
2. - I'll have to think on this. While I agree it would be a fun feature, it would need to be very, very strict just to keep the amount of code to a minimum and to reduce potential attack vectors from a security perspective. I may just steal some of the formatting functionality I added to Theater of the Mind.
    - This shouldn't pose a problem at all. The hardest part would be deciding on rarity levels and then figuring out a way to implement them on the UI side of things without making it too cluttered.  A dropdown with basic rarity options would be ideal.
     - I favor the tag system and again, while easy to implement, the UI/UX side of things will be more tricky. How would we handle adding and removing tags in a way that didn't clutter the UI? What about editing a tag? Etc.
3.  Other than offering the ability to view achievements with specific tags (then you could use them for categorization and sorting) what did you have in mind?

**Comment by andrey-pavlov-13** [comment]
> I'll have to think on this. While I agree it would be a fun feature, it would need to be very, very strict just to keep the amount of code to a minimum and to reduce potential attack vectors from a security perspective. I may just steal some of the formatting functionality I added to Theater of the Mind.

Well, possibility to make paragraphs it`s enough for me to be honest. But a wider ability to edit text could be really yammy.

> This shouldn't pose a problem at all. The hardest part would be deciding on rarity levels and then figuring out a way to implement them on the UI side of things without making it too cluttered. A dropdown with basic rarity options would be ideal.

I think, it`s possible to add this as tags to each item. Actually this one is really tags-themed stuff in my opinion.

> I favor the tag system and again, while easy to implement, the UI/UX side of things will be more tricky. How would we handle adding and removing tags in a way that didn't clutter the UI? What about editing a tag? Etc.

What about add one additional field where you can add tags with simple comma. After editing it could looks like item properties in D&D/PF2e.


> Other than offering the ability to view achievements with specific tags (then you could use them for categorization and sorting) what did you have in mind?

I mean about the way to change position for items like the way you can sort Notes in Foundry: alphabetically or just manual sorting, when you decide where each item should be.

**Comment by EddieDover** [comment]
> > I'll have to think on this. While I agree it would be a fun feature, it would need to be very, very strict just to keep the amount of code to a minimum and to reduce potential attack vectors from a security perspective. I may just steal some of the formatting functionality I added to Theater of the Mind.
> 
> Well, possibility to make paragraphs it`s enough for me to be honest. But a wider ability to edit text could be really yammy.

I think line breaks, italics, and bold should be easy enough, so those will probably happen first.

> 
> > This shouldn't pose a problem at all. The hardest part would be deciding on rarity levels and then figuring out a way to implement them on the UI side of things without making it too cluttered. A dropdown with basic rarity options would be ideal.
> 
> I think, it`s possible to add this as tags to each item. Actually this one is really tags-themed stuff in my opinion.

I'm not sure the tag route would work here. You would still need a way to match a tag to a color and then handle things if the tags were changed.  Maybe agreeing on a rarity list similar to the MMO standard would be best. Common to Legendary, that sort of thing.

> 
> > I favor the tag system and again, while easy to implement, the UI/UX side of things will be more tricky. How would we handle adding and removing tags in a way that didn't clutter the UI? What about editing a tag? Etc.
> 
> What about add one additional field where you can add tags with simple comma. After editing it could looks like item properties in D&D/PF2e.

I was thinking either that approach, or a self-search combo select.  Honestly the former would be less code so that might be the way to go.

> 
> > Other than offering the ability to view achievements with specific tags (then you could use them for categorization and sorting) what did you have in mind?
> 
> I mean about the way to change position for items like the way you can sort Notes in Foundry: alphabetically or just manual sorting, when you decide where each item should be.

![image](https://github.com/EddieDover/fvtt-player-achievements/assets/464286/f3676b44-9c6a-42b3-813e-14aba387a815)
This button should already allow you to sort alphabetically A-Z or Z-A (however I just noticed it might not be working correctly for some reason...)

Manual sorting would be a pain to implement, and moreover, having the system save your previously sorted items and recall them between gameplay sessions, could be a pain. I'll have to think on it.

**Reply by andrey-pavlov-13** [reply]
> I think line breaks, italics, and bold should be easy enough, so those will probably happen first.

That's more then enough for me tbh.

> Manual sorting would be a pain to implement, and moreover, having the system save your previously sorted items and recall them between gameplay sessions, could be a pain. I'll have to think on it.

I just give some ideas which could improve module. You are free to ignore any of it if you think that it is impossible or too hard to implement by some reasons.

**Comment by EddieDover** [comment]
@andrey-pavlov-13 Thoughts on #52 ?

**Comment by EddieDover** [comment]
@andrey-pavlov-13 Also, please take a look at #54

**Reply by andrey-pavlov-13** [reply]
This one is really great. That`s what i talking about.

**Comment by EddieDover** [comment]
Closing this discussion for now, just to keep things clean.  If you'd like to suggest further adjustments or remind me of ones I forgot, please do so in another discussion.  I'll also see about opening a Trello board or some place where people can vote on things they'd like me to focus on.
